### ISSUE TITLE
SSTable CLI tools no longer work in 5.4 - Throws NumberFormatException

### ISSUE THREAD

**Issue by fee-mendes** [scylladb/scylladb#16339]
**Installation details**
Scylla version (or git commit hash): `5.4.0-0.20231205.58a89e7a4231` with build-id `7486a70ced30f7c89053f5c5a230c20ad31ffc1f`

**Description**

Trying to inspect inner SSTable contents fails with a `NumberFormatException`, as it seems we are sending the UUID identifier instead. Related to https://github.com/scylladb/scylladb/commit/d1dc579062fb599e0e0cee75e8592f17b2a1884d ?

```shell
root@df13a4d04afa:/var/lib/scylla/data/foo/bar-eb90daf0952d11ee87e220c83109c794# sstablemetadata 'me-3gbp_1glu_4e6g020ns4px173el0-big-Data.db'
Exception in thread "main" java.lang.NumberFormatException: For input string: "3gbp_1glu_4e6g020ns4px173el0"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at org.apache.cassandra.io.sstable.Descriptor.fromFilename(Descriptor.java:280)
	at org.apache.cassandra.io.sstable.Descriptor.fromFilename(Descriptor.java:228)
	at org.apache.cassandra.io.sstable.Descriptor.fromFilename(Descriptor.java:217)
	at org.apache.cassandra.tools.SSTableMetadataViewer.main(SSTableMetadataViewer.java:150)
root@df13a4d04afa:/var/lib/scylla/data/foo/bar-eb90daf0952d11ee87e220c83109c794# sstabledump 'me-3gbp_1glu_4e6g020ns4px173el0-big-Data.db'
Exception in thread "main" java.lang.NumberFormatException: For input string: "3gbp_1glu_4e6g020ns4px173el0"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at org.apache.cassandra.io.sstable.Descriptor.fromFilename(Descriptor.java:280)
	at org.apache.cassandra.io.sstable.Descriptor.fromFilename(Descriptor.java:228)
	at org.apache.cassandra.io.sstable.Descriptor.fromFilename(Descriptor.java:217)
	at org.apache.cassandra.tools.SSTableExport.run(SSTableExport.java:188)
	at com.scylladb.tools.SSTableExport.main(SSTableExport.java:44)
```

**Comment by tchaikov** [comment]
@mykaul and @fee-mendes  the java tools do not support the sstables with uuid-based identifiers. see https://github.com/scylladb/scylla-tools-java/issues/333 . the plan is to encourage our users and field engineers to use `scylla sstables` . see https://opensource.docs.scylladb.com/stable/operating-scylla/admin-tools/sstablemetadata.html

@fee-mendes are the `scylla sstable` commands able to offer what we need? or do we have any compelling reasons that we need to continue using `sstablemetadata` in newer Scylla v5.4 and up ?

**Comment by tchaikov** [comment]
> > @fee-mendes are the scylla sstable commands able to offer what we need? or do we have any compelling reasons that we need to continue using sstablemetadata in newer Scylla v5.4 and up ?
> 
> I don't know the answer, but taking a quick look at its documentation it does seem to cover (even more than) what we need. I assume we'll need to get used to its output though.
> 
> But really, @tchaikov the documentation you linked claims we are deprecating it and we recommend users to start getting familiar with the new tooling. Here we are actually forcing people to switch with no further notice.

practically, yes. despite that we do have an option `uuid_sstable_identifiers_enabled` to prevent scylla from using uuid-based identifiers, it is enabled by default. so under the most circumstances, once user installs/upgrades to 5.4, they are dropped in the era of sstables with these uuid names.

> 
> Even then, wouldn't it be better to simply sym/hardlink the old commands to the new one so that users actually don't need to figure this out?

these java-based tools have different command line options, and their output formats are different from their corresponding `scylla sstables` commands, we cannot just replace them with symlink. it would bring more confusion, i think. 

also, because these tools can still be used with sstables named with the integer-based generation. they are not removed at this moment, even though we can.

> 
> Unrelated to this issue, but something I just realized... one of the primary motivations for inspecting SSTable contents would be during an investigation involving a particular shard (ie: large tombstone run), but the new UUID based format kinda complicated things as computing the shard_id is no longer intuitive (gen % smp_count).

i understand why it's nice to have, as it does help us to pinpoint performance issues. please see the reply below.

> 
> Rather, we need to std::hash it, which [doesn't strictly specify a hash function](https://stackoverflow.com/a/12623446). So I can see how the use cases are now somehow limited in the field, and likely we should have a separate issue for adding support to print a shard ID given a specific SSTable UUID generation

@denesb probably we can add this to one of `scylla sstable` commands? or add yet another command? as all existing commands are designed around the content of sstables: they either read or mutate the components of specified sstable(s). but the connection between a certain sstable's identifier and the shard is quite a different knowledge to expose. i recall that you and @raphaelsc kept reminding me that the connection is not solely determined the identifier when i was working on the uuid-identifier change. hence this comment:
```c++
    /// returns a hint indicating if an sstable belongs to a shard. The definitive
    /// way to determine that is overlapping its partition-ranges with the shard's
    /// owned ranges.
```
what do you think?

**Comment by denesb** [comment]
> @denesb probably we can add this to one of `scylla sstable` commands? or add yet another command? as all existing commands are designed around the content of sstables: they either read or mutate the components of specified sstable(s). but the connection between a certain sstable's identifier and the shard is quite a different knowledge to expose. i recall that you and @raphaelsc kept reminding me that the connection is not solely determined the identifier when i was working on the uuid-identifier change. hence this comment:
> 
> ```c++
>     /// returns a hint indicating if an sstable belongs to a shard. The definitive
>     /// way to determine that is overlapping its partition-ranges with the shard's
>     /// owned ranges.
> ```
> 
> what do you think?

This is definitely possible. We can call said command `scylla sstable shardof` and it would print out the intersection between the shard-ranges and the partition ranges (which should usually be a single shard).

**Comment by denesb** [comment]
@tchaikov maybe we should patch these java-based tools to print a message when they see a UUID generation, telling the user to user the native tools for sstables with UUID generation.

**Comment by tchaikov** [comment]
thanks for the prompt reply. created #16343

**Comment by tchaikov** [comment]
> @tchaikov maybe we should patch these java-based tools to print a message when they see a UUID generation, telling the user to use the native tools for sstables with UUID generation.

thank you for the suggestion. it's better than removing them or adding a symlink to `scylla`. lemme update https://github.com/scylladb/scylladb/issues/14856 to add this as a work item.

**Comment by mykaul** [comment]
This on is actually fixed via https://github.com/scylladb/scylla-tools-java/pull/362 - we just need a backport.

**Comment by denesb** [comment]
> This on is actually fixed via [scylladb/scylla-tools-java#362](https://github.com/scylladb/scylla-tools-java/pull/362) - we just need a backport.

Already backported as 9fc4c265a56ce04ea52d0e5bb2d570738d9f9f91, but wasn't promoted yet.
